test/functional/terminal/ex_terminal_spec.lua

```diff
@@ -48,6 +49,39 @@ describe(':terminal', function()
     eq(2, eval("line('.')"))  -- cursor stays where we put it
   end)
 
+  it('does not crash when callback fires after terminal closes', function()
+    screen:detach()
+    screen = thelpers.screen_setup(0, '["'..helpers.nvim_prog..'", "-u", "NONE", "-i", "NONE", "-N"]')
+    thelpers.feed_data({':let g:split = 0',
+                        ':au TermOpen * nnoremap <buffer> x :q<CR>',
+                        ':nnoremap <silent> x :new <bar> let g:split += 1 <bar> call termopen([&shell])<CR>',
+                        ':set updatetime=0',
+                        ''})
+    screen:expect([[
+      {1: }                                                 |
+      {4:~                                                 }|
+      {4:~                                                 }|
+      {4:~                                                 }|
+      {5:[No Name]                                         }|
+      :set updatetime=0                                 |
+      {3:-- TERMINAL --}                                    |
+    ]])
+    local count = 500
+    for _ = 1,count do
+      thelpers.feed_data('xx')
+      wait()
+    end
+    thelpers.feed_data({':echo g:split', ''})
+    screen:expect([[
+      {1: }                                                 |
+      {4:~                                                 }|
+      {4:~                                                 }|
+      {4:~                                                 }|
+      {5:[No Name]                                         }|
+      ]]..tostring(count)..string.rep(' ', 50 - string.len(tostring(count)))..[[|
+      {3:-- TERMINAL --}                                    |
```

Comment:
> We mostly just want to give ASAN a chance. So maybe 10 iterations instead of 500 is good enough.

The conditions where it happens seem to be when there's a backlog of input being processed, so the queue is running faster than the terminal/shell can startup and teardown.  When I manually reproduce it, I stop pressing the x and nvim keeps chugging along for a few more seconds starting/killing terminals before it crashes.

If there's a better way to simulate that stress to increases the chances of hitting the problem, I'm all ears.  If we're not that concerned with exactly hitting it and just want some coverage, then I'm also fine with lowering the counts and simplifying the logic.